agreement/pseudonode.go

```diff
@@ -193,42 +195,61 @@ func (n asyncPseudonode) MakeVotes(ctx context.Context, r round, p period, s ste
 	}
 }
 
-func (n asyncPseudonode) makeProposalsTask(ctx context.Context, r round, p period) pseudonodeProposalsTask {
-	participation := n.getParticipations("asyncPseudonode.makeProposalsTask", r)
+// load the participation keys from the account manager ( as needed ) for the
+// current round.
+func (n *asyncPseudonode) loadRoundParticipationKeys(voteRound basics.Round) []account.Participation {
+	// if we've already loaded up the keys, then just skip loading them.
+	if n.participationKeysRound == voteRound {
+		return n.participationKeys
+	}
 
+	cparams, err := n.ledger.ConsensusParams(ParamsRound(voteRound))
+	if err != nil {
+		// if we cannot figure out the balance round number, reset the parameters so that we won't be sending
+		// any vote.
+		n.log.Warnf("asyncPseudonode: unable to retrieve consensus parameters for voting round %d : %v", voteRound, err)
+		n.participationKeysRound = basics.Round(0)
+		n.participationKeys = nil
+		return nil
+	}
+	balanceRound := balanceRound(voteRound, cparams)
+
+	// otherwise, we want to load the participation keys.
+	n.participationKeys = n.keys.VotingKeys(voteRound, balanceRound)
+	n.participationKeysRound = voteRound
+	return n.participationKeys
+}
+
+func (n asyncPseudonode) makeProposalsTask(ctx context.Context, r round, p period) pseudonodeProposalsTask {
 	pt := pseudonodeProposalsTask{
 		pseudonodeBaseTask: pseudonodeBaseTask{
-			node:          &n,
-			context:       ctx,
-			participation: participation,
-			out:           make(chan externalEvent),
+			node:    &n,
+			context: ctx,
+			out:     make(chan externalEvent),
 		},
 		round:  r,
 		period: p,
 	}
-	if len(participation) == 0 {
+	if !pt.populateParticipationKeys(r) {
```

Comment:
It's a "functional programming" approach to avoid side effects, so this misses the point a little bit by providing a similar guarantee using a side effect. To your point earlier, this is a private function that isn't used widely, so I don't think this nit is important enough to worry about further.